File: client/web/src/shell/nav-alarm.ts
import type { Band } from "../screens/questions/contract";
import { bandTone } from "../screens/status-board/tile-copy";

// The Status nav control's tint, and the one thing this layer is allowed to
// decide about it.
//
// **Which band came back is the core's answer** (`decisions/panes/alarm.rs`,
// reached through `statusAlarm`), not this file's: the two opinions that
// fold makes — a gap is silent, `dormant` is silent, everything else is not
// — are the same on the phone, the desktop and Android, so they live once,
// in Rust. What is left here is the last step alone, the step ADR-0025
// leaves per-client: which colour a band paints as.
//
// **And even that is not a new mapping.** `bandTone` is the board's own
// band→treatment table, imported rather than restated, so the button and the
// tiles it expands into can never disagree about how bad the same answer is.
// The direction of the import (shell reading from a screen) is the price of
// having exactly one table; the alternative was a second copy in `shell/`,
// which is the thing that would actually rot.

/** The colour the Status control wears for `band`, or `undefined` when it
 * should wear its ordinary nav colour.
 *
 * `undefined` in, `undefined` out — "nothing raises the nav" and "this band
 * is quiet" are the same instruction to a caller, and splitting them would
 * make every call site handle two spellings of one answer. */
export function navAlarmColor(band: Band | undefined): string | undefined {
  if (band === undefined) return undefined;
  switch (bandTone(band)) {
    case "danger":
      return "var(--status-danger-fg)";
    case "warn":
      return "var(--status-warn-fg)";
    // `dormant` reaches here only if the core ever starts returning it; the
    // nav wears its ordinary colour rather than inventing a third tint.
    default:
      return undefined;
  }
}

File: client/android/app/src/main/kotlin/net/twinion/hummingbird/NavAlarm.kt
package net.twinion.hummingbird

import androidx.compose.ui.graphics.Color
import net.twinion.hummingbird.ui.theme.Amber600
import net.twinion.hummingbird.ui.theme.Crimson600
import net.twinion.hummingbird.ui.theme.CrimsonDark
import net.twinion.hummingbird.ui.theme.StatusWarnFgDark
import uniffi.hummingbird_ffi_mobile.MobilePaneBand

/** The Status nav destination's tint, and the one thing this side decides
 * about it — the Android half of the web's `shell/nav-alarm.ts`.
 *
 * **Which band came back is the core's answer**
 * (`decisions/panes/alarm.rs`, reached through `MobileTaskHost.statusAlarm`),
 * not this file's: whether a gap counts and whether `dormant` counts are the
 * same questions on the phone, the tablet and both web nav forms, so they
 * are answered once, in Rust. What is left here is the last step ADR-0025
 * leaves per-client: which colour a band paints as.
 *
 * The band→tone step is the board's own, ported rather than re-decided —
 * `tile-copy.ts`'s `bandTone`, which the status tiles already apply to these
 * same answers: live/imminent read as danger, near/distant as warn. Keeping
 * the two identical is what stops the button and the tiles it opens onto
 * disagreeing about how bad one answer is.
 *
 * **Plain, not `@Composable`, and `dark` is a parameter.** `MaterialTheme
 * .colorScheme.error` already carries `--status-danger-fg` in both scopes
 * (`Theme.kt`) and reading it would have been the shorter route, at the
 * price of making this whole mapping renderable-only — a Robolectric
 * harness to assert a colour. Taking the two token pairs by name instead
 * leaves the mapping a pure function with a JVM test, and the pairs are the
 * ones `syncStatusToneColor` already reads for the same two tones.
 *
 * `null` in, `null` out: "nothing raises the nav" and "this band is quiet"
 * are one instruction to a caller, and splitting them would make every call
 * site handle two spellings of one answer. */
internal fun navAlarmColor(band: MobilePaneBand?, dark: Boolean): Color? = when (band) {
    null -> null
    // `--status-danger-fg`: `var(--crimson-600)` light, its own literal dark.
    MobilePaneBand.LIVE, MobilePaneBand.IMMINENT -> if (dark) CrimsonDark else Crimson600
    // `--status-warn-fg`: `var(--amber-600)` light, its own literal dark —
    // NOT `UrgencySoonDark`, a different token that happens to share the
    // dark value and would drift the day either moves.
    MobilePaneBand.NEAR, MobilePaneBand.DISTANT -> if (dark) StatusWarnFgDark else Amber600
    // The core folds `dormant` away before it ever reaches here; a nav that
    // invented a third tint for it would be painting "everything is fine"
    // as a state worth noticing.
    MobilePaneBand.DORMANT -> null
}
